### ISSUE TITLE
plan for next major release

### ISSUE THREAD

**Issue by al-caveman** [al-caveman/nsapass#6]
hi - so `nsapass v0` is nice, works, < 500 lines of code, easy to audit, no segfaults, blah blah blah.  but imo there is room to be even more lazy.  here we discuss the plan for the next major release.

# simplicity
lines should remain within 500 to allow for convenient auditing.

# syntax
`nsapass get caveman` is too much typing imo.  imo we can do much better.

- `get` must be implied.  e.g. it should be `nsapass caveman`.
- and syntax must be `nsapass QUERY`, where `QUERY` is doesn't require full match.  
  - e.g. `nsapass c` should automatically resolve to `nsapass caveman` if `caveman` is the only username that starts with letter `c`. 
  - but if we have multiple entries that start with `c`, then `nsapass c` should prompt us:
```
4 entries match your query "c":
    0. caveman@protonmail.com
    1. caveman@freenode.net
    2. caveman@dal.net
    3. cavetest@stackexchange.com

choose entry [0-3]: 
     
```

**so here is a fundamental question:** what should be the language of `QUERY`?

here is another syntax:
```
nsaadd [-u USERNAME] [-p PASS] [-l LEN] [-s SEST] [-c COMMENT]
nsaget QUERY
nsadel ...
nsals ... (ls short for less)
nsamod ...
nsachangepass
```

where all these "commands" are nothing but symlinks to `nsapass` executable.  so the project will remain a single small py file, except that during the installation, it will get symlinks `nsadd`, `nsaget`, ..., linked to it.

then advatages would be:

- less typing since we removed `pass `.
- user can get list of commands by tabs completion.  e.g. `nsa<TAB TAB>` will show all sub-commands in shell.
- plus, as a bonous side, it will be more funny?  as it literally says that we are adding/getting our passwords from nsa, directly.

but disadvantages:

- more complex installation as user would need to install symlinks.

is it worth it?  so far i think _"yes worth it"_ but not sure (i feel i am mayb missing something).

# db structure
currently it's like this:

```json
{
    "entries" : 
        {
            "caveman" : 
                [
                    {"pass":"******", "note":"for freenode"},
                    {"pass":"******", "note":"for dalnet"},
                    {"pass":"******", "note":"for protonmail"},
                ],
            "cavetest" :
                [
                    {"pass":"******", "note":"for stackexchange"},
                ],
        }

    , "archived":
        {
            "caveman" : 
                [
                    {"pass":"******", "note":"old freenode password"},
                ],
        }
}
```

but several issues:

- is `archived` the right name?  should it be called "trashbin"?  because it's basically a trashbin.
- should we even have a trashbin?
- should `note` be a notes section?  or should it be a tags section?  because depending on how we define `QUERY`, we may need to search inside `note`.
- the whole tree structure is questionable.  people usually don't have too many gigs of passwords.  so the structure must not be optimized for speed because it's not useful in this case.  instead, the structure better be optimized for flexibility in `QUERY` to allow for fastest identification of desired entry.

## alt 1
here is an alternative structure:
```json
[
    {"username":"caveman", "pass":"*****", "comment":"for freenode", "deleted":false},
    {"username":"caveman", "pass":"*****", "comment":"for dalnet", "deleted":false},
    {"username":"caveman", "pass":"*****", "comment":"for github", "deleted":false},
    {"username":"cavetest", "pass":"*****", "comment":"for stackexchange", "deleted":true},
    .
    .
    .
    etc
]

```

with this approach, `nsapass` will need to search throughout the whole list to find the matching entries based on `QUERY`.  its complexity will be `O(n)`, but `n` is small in reality for this application, so it might be the best structure to allow for the most convenient searching by `QUERY`.

since we are storing the username in its own field, should we have the feature to load it into clipboar too?  e.g. upon loading an entry, we copy the username in a buffer, and the password in another buffer, so that the user pastes both into whatever form fields he wants.  to generalize this, we may replace:
```
CLIPCMD = ['xclip', '-selection', 'clipboard']
```
by an indefinite list of copy commands:
```
COPYCMDS = [
    {"command": ['xclip', '-selection', 'clipboard'], "stdin":"{PASSWORD}"}, 
    {"command": ['xclip', '-selection', 'primary'], "stdin":"{USERNAME}"}, 
]
```

but that seems a bit hask-ish, no?  any better alternative to copy the username and password?

## alt 2
or.. how about this (no `username`):
```json
[
    {"pass":"*****", "comment":"caveman @ freenode", "deleted":false},
    {"pass":"*****", "comment":"caveman @ dalnet", "deleted":false},
    {"pass":"*****", "comment":"caveman @ github", "deleted":false},
    {"pass":"*****", "comment":"cavetest @ stackexchange", "deleted":true},
    .
    .
    .
    etc
]

```

since we don't load the username in a clipboard so far, what is the value of storing it in a field on its own?  why shouldn't it be in the comments section?

# more?
please do tell!

**Comment by int-e** [comment]
There's some UI discussion in #5 that better fits here:

We should try to make the UI consistent, so let's look at all commands at once. The `NAME INDEX` part (with the index being optional for `get` because it is not destructive) is applicable to almost all commands, so it should be part of positional arguments.

Here is a list for discussion:
```
usage: nsapass get [-h] NAME [INDEX]
usage: nsapass add [-h] [-P | -p PASS | [-l LEN] [-s SET]] [-c NOTE] NAME
usage: nsapass del [-h] NAME INDEX
usage: nsapass undel [-h] NAME INDEX
usage: nsapass mod [-h] [-n NEWNAME] [-P | -p NEWPASS] [-c NEWNOTE] NAME INDEX
usage: nsapass list [-h] [-a]
usage: nsapass changepass [-h]
```
This assumes that there are some sensible defaults for `-l` and `-s`.

The `INDEX` could be made optional even for destructive commands when there's only one possible index.

The `mod` command could also take `-l` and `-s` options.

**Comment by al-caveman** [comment]
ya.  good points, ty.  

so i sat down, and tried to re-think the whole thing, and came up with this preliminary suggestion:

```
usage: nsapass get          [-h] [QUERY] [BEHAVIOR]
usage: nsapass add          [-h] [BEHAVIOR] [-u USERNAME] [-i INDEX] [PASSWORD] [-n NOTE]
usage: nsapass del          [-h] [QUERY] [BEHAVIOR]
usage: nsapass undel        [-h] [QUERY] [BEHAVIOR]
usage: nsapass mod          [-h] [QUERY] [BEHAVIOR] [-u USERNAME] [-i INDEX] [PASSWORD] [-n NOTE]
usage: nsapass list         [-h] [-a]
usage: nsapass changepass   [-h]

BEHAVIOR = [-p | -b]
QUERY = USERNAME [INDEX]
PASSWORD = [-k KEY | -l LEN -s SET]
```

which behaves as follows:

- `BEHAVIOR` specifies whether ambiguities are resolved interactively via prompts (i.e. if `-p` is given), or whether they are not resolved and —instead— produce an error and exit (i.e. batch mode if `-b` is given).
- `INDEX` is always guessed if not given.  if no sensible guess is available, then this ambiguity is resolved by the value of `BEHAVIOR`.
- `PASSWORD` is either given via cli (i.e. `-k KEY`) or generated automatically (i.e. `-l LEN -s SET`).  if none are given, then such ambiguity is resolved based on `BEHAVIOR`.
- `get` as expected.  if no `BEHAVIOR` is specified, `-p` is assumed.
- `add` as expected.  if any of the optional arguments are missing, then it will prompt the user interactively for whatever not specified.  if no `BEHAVIOR`,  '-p' is assumed.
- `del` as expected.  if no `BEHAVIOR`, '-p' is assumed.
- `undel` as expected. if no `BEHAVIOR`, '-p' is assumed.
- `mod` as expected. **if no `BEHAVIOR`, *'-b'* is assumed**.
- `list` as expected.
- `changepass` as expected.

also  no `NEWX` in `mod` because technically it's value has the same specification as `X`.

this also assumes that we'll change variable names, since "name" is now "username", for consistency with arguments.

**Comment by al-caveman** [comment]
extra thought:  should we drop the archive feature?  it will simplify the definition of `INDEX`, and stops re-inventing backup solutions.

but not sure if the risk is too high for accidentally  deleting unbacked up passwords.

**Comment by int-e** [comment]
No real opinion yet (sorry, this may take a day or two), just a couple of further thoughts:

* how does `mod` interact with a typical website password change? typically you'll need
  1. the original password, to log in
  2. the new password (often twice, not that it really matters)
  3. often the original password again, to confirm (this is the awkward bit if you are not careful, especially if the archive is removed)
* when generating new passwords, it would be nice to have an option to never display them, but exclusively go through clipboard instead
* should there be a command for copying entries?
* one opinion: I'd be sad to lose `add [NAME]`

**Comment by al-caveman** [comment]
> how does mod interact with a typical website password change?

here is how i guess most people will use `mod`:

1. user runs `snapass mod mynick -p`.  (hmm.. i guess `-p` better be default here, too, so we don't type it.  so it will be just `snapass mod mynick`, or if we incorporate the symlink idea `nsamod mynick`, or if we incorporate smarter query by minimum match then just `nsamod m` in case `mynick` is the only username that starts with `m`.)
2. user enters password to decrypt db.
3. user gets prompted, with default options being previous value.  he hits enter key to advance to next prompt.  since he doesn't want to change comments, username, etc, he will just hit enter enter enter until he reaches the password section.
4. when the user reaches the password-change prompt, he will be offered a newly generated password based on default `LEN` and default `SET`.  the prompt will allow him to accept it, get another generated password (possibly with other `LEN` and `SET` values), or supply his own `KEY`.
5. when happy, the user is shown a summary of the change to be permanently saved in db.
6. the user puts the password twice to save the db.

yes, stage (6) is not elegant.  i don't know how to fix it.  so currently i left it like this and used it as a safeguarding feature to ensure that a user doesn't commit a change to the database accidently (him typing password twice is some kind of confirmation that he indeed is happy about the changes made).

or, if a user wants to automate it, he may run it in batch mode:
```
nsapass mod mynick -l 30 -s printable -b
```
then, automatically, the password is replaced by a newly randomly generated password.

> when generating new passwords, it would be nice to have an option to never display them, but exclusively go through clipboard instead

good point.

> should there be a command for copying entries?

probably, but not sure. i think if we stumble upon scenarious in our daily life where we feel that a copy would've made our life easier, then we will add a copy feature.  however, now, i personally didn't stumble upon this case yet, so i didn't add it purely in the spirit of occam's razor.  so i'm currently waiting to stumble upon this scenario, then add it.

i mostly fooled myself with this "safeguarding" feature so that my ocd doesn't eat me.  if you know of a solution i'd like to know.

if you have stumbled upon it, then it is justified and should be added.

> one opinion: I'd be sad to lose add [NAME]

how about this: 

```
usage: nsapass add          [-h] [QUERY] [BEHAVIOR] [PASSWORD] [-n NOTE]
```

?

**Comment by int-e** [comment]
> here is how i guess most people will use mod:

Ah but I asked about how this *interacts* with a password change on a website, not how it compares to that.

As I see it, there will be two simultaneous interactions: One interaction with the website (to update the password there), and another one with `nsapass mod` to generate and update the password in its db. When step 5 in the `mod` process is reached, two things can still happen on the website end:

* we may be prompted for the original password. This may be hard if we've already completed the `mod` process, unless we've kept a backup.
* worse, the website may reject the generated password, so we'll want to undo the change `mod` did.

Maybe the whole idea of using `mod` to generate a new password is flawed... and people should instead add a new password, query both the old and new one as needed, and then delete the old password once the update is accomplished on the website? I'm not sure.

**Comment by al-caveman** [comment]
oo.  i see.  sry i missed that part.  you shed light on a great issue.  awesome help! thanks you a lot.

## blah blah

personally, my plan is to run `mod` only after the site prompts me with `new password`.  so i login normally with previous passwords, and only run `mod` after i get the `new password` prompts from the site.

this is how i was doing it with `keepassxc` too.  they have a botton where i press to regenerate a new password.  i used to use that botton.

but the fact remains that this process is very risky and dirty.  the workflow of passwords changes need more thinking.  do you have any better ideas?

the good news with `nsapass` is that `mod` is just a wraper around `add` and `del`, so old password is archived just in case we need to get back to it.  so i think this aspects makes `nsapass`'s process safer than how i used to use `keepassxc` since password updates there are not archived (only deletes are are archived, which is funny imo).

## questions

any idea how other password managers solve this?  and how do they do a better job in this regard?

maybe i'm missing something fundamental?  because everything i did is based on _my_ use of `keepassxc`. i don't know how others use their passwords managers.

**Comment by al-caveman** [comment]
https://github.com/Al-Caveman/nsapass/pull/5#issuecomment-663607576

**Comment by int-e** [comment]
I would still prefer `add` to not use `-u` nor `-i`. Other than that I like where this is going.

The most problematic feature of the database design is that indices are bound to be confusing, *if one actually relies on them*.

However, if one sticks to the principle that every user (= site?) should have a single entry, with the exception of going through a password change and possibly keeping the old password around, then indices should be perfectly okay.

So maybe the best way forward is to document a suggested workflow that avoids indices most of the time.

* make one username per site+login combination, use `add` for initial entry
* use `mod` to generating a new password, rely on the archived version to query the old one if it turns out to be needed
* use `del` to get rid of the archived version after the password update is completed

This ensures that any user will have at most 2 used indices, and only one that is not archived, so you will not have to worry about indices 99% of the time.

This is just my view of course. If you have more elaborate uses for indices in mind this may require some hard thought.

**Comment by al-caveman** [comment]
ya.  the indices part is true.  i thought about it, but then i opted against it because i asked myself `do i need this restriction?`.  the answer to that is imo `no`.  so i thought, then why add this restriction?

now, this discussion made me think more to re-consider another thought:  how about named indices where we call them `HOST`?  this way a query statement will be like `nsapass get USERNAME HOST`.

the advantage of the named indices is that we will get to apply the shortest-match-to-uniqueness principle already applied in `USERNAME` to named indices (aka `HOST`).  e.g. `nsapass get c p` will automatically understand that i mean `caveman` at the host `protonmail.com`.

but if we don't use named indices, then a `NAME` would need to be longer to contain the host part, such as `caveman@protonmail.com`, which will complicate the matching if there is `caveman@hotmail.com`, since the shortest query to identify a unique match will be `nsapass get caveman@p`.

as for `USERNAME` `HOST` for `add`, i agree.  my plan is to add them, but i just  kept it out to let the basic idea settle in my head 1st (specially that i rewrote the argparse bit).

**Comment by al-caveman** [comment]
a bit unrelated, but just to make sure i don't forget:  i suggest, after we do what's needed to close this issue (and related prs), to make it `nsapass v5.0.0` instead of `nsapass v1.0.0`.  still sort of semver, just major version bumped multiple steps, large enough to make me feel happy.

`nsapass v3.0.0` is not enough imo.  it's boring.  but `nsapass v5.0.0` _feels_ good.  dunno what do you think.  i hope you share the same feeling?  or do you have even _better_ feelings?

as for the release after, depending on what happens, it may suddenly shoot into `nsapass v9000.0.0` (and that would be 3rd version).

at this rate of unprecedented awsomeness, i wonder if gentoo can handle us?  gentoo thinks largest version is `9999` i think.  you think we can beat it in 4th release? 😆

**Comment by int-e** [comment]
For addressing entries, one could simply have an 'ID' field with no assigned semantics... then users can pick whatever naming scheme they want. (In fact, I was going to use 'NAME' in such a manner. I'd use the NOTE field to record account names and site if not encoded in the name...)

I guess a lot depends on how many accounts you expect to manage and whether you're in the habit of having multiple accounts per site. It may be hard to fit all uses at once.

As for versions, to the extent that I care (which is very little) I'd be boring and go with 1.0.0. Starting with a high version number may seem fun, but it's hard to impossible to go back if one later regrets it.

**Comment by al-caveman** [comment]
should we get rid of indices?  and require users to select unique `ID` fields?  e.g. give warning `ID exists` (if we have revisions in fields)

**Comment by int-e** [comment]
Hmm, okay, in the interest of making a (mostly) definitive proposal, yes, let's get rid of the indices. But let's keep the argument itself for a revision number, where 0 is the latest revision, 1 the previous one, and so on. So we would have:

* ```nsapass get [-h] [-z] [ID] [REV]```
  Get entry. REV defaults to 0, but can be used to access previous revisions.
* ```nsapass add [-h] [-z] [-g LEN SET | -m] [-n NOTE] [ID]```
  Add entry. Existing entries are moved by 1, thus becoming older revisions.
* ```nsapass del [-h] [-a] [ID] [REV]```
  Remove a revision. `-a`: remove all revisions associated with the [ID]. Cannot be undone, except for REV=0 if the refinement below is implemented.
  NB why does `del` currently have a `-z` flag?
* ```nsapass undel [-h] [ID] [REV]```
  Undelete. Basically copies (or moves?... choices, choices) an old revision to the front of the list.
  If it copies, this will be equivalent to `nsapass mod [-h] [ID] [REV]`.
  NB why does `undel` currently have a `-z` flag?
* ```nsapass mod [-h] [-z] [-g LEN SET | -m] [-i ID] [-n NOTE] [ID] [REV]```
  Modify an entry. Adds a new revision, possibly under a new name, in which case it copies an entry. Can access old revisions. Does not delete anything.
* ```nsapass view [-h] [-o] [-z] [ID]?```
  View entries. `-o` = include old revisions. `[ID]` could be added to filter entries by a prefix
* ```nsapass chpass [-h]```
  Unchanged.

Refinement: In order to allow deletion of the latest revision *with* undo, one could allow the latest revision to be `None`. Then `del ID` could just add a `None` in front, and `undel` could revert that.

*Note* that this is very similar to the existing interface, but by treating the indices as revisions and removing the ability to add new entries at arbitrary indices (added entries will always be revision 0 of their respective ID), the shifting of indices (now revisions) becomes a feature rather than a constant source of confusion. Note also that the archival flag will disappear; instead, all revisions with index larger than 0 will behave like archived ones.

**Comment by al-caveman** [comment]
any idea what to store in revisions?  e.g. should `ID` renames cause revisions?

**Comment by int-e** [comment]
`ID` renames are not currently on the list... `mod` can take an entry from one `ID` and add it (with modifications) to another `ID` (which may or may not already exist). If `ID` renames are added (which does seem reasonable) I imagine that would just move the revisions to the new ID with no backup.

**Comment by al-caveman** [comment]
yeah.

`/me thinking loudly`:

```
{
    "caveman protonmail": [  <---------------+
        {                                    |-- same
            "id": "caveman protonmail",   <--+
            "password": "*******",
            "note": "latest note",
            "timestamp": 1594447546.9914703,
            "archived": false
        },
        {
            "id": "urbanman hotmail",
            "password": "*******",
            "note": "old note",
            "timestamp": 1593337546.9914703,
            "archived": false
        },
        {
            "id": "18-m-AF aol",
            "password": "*******",
            "note": "oldest note".
            "timestamp": 1592227546.9914703,
            "archived": false
        }
    ],
    "some other account": [  <-------------+
        {                                  |-- same
            "id": "some other account",  <-+
            "password": "*******",
            "note": "latest note",
            "timestamp": 1591117546.9914703,
            "archived": true   <---- deleted/archived.
        },
        {
            "id": "some other account",
            "password": "*******",
            "note": "latest note",
            "timestamp": 1590007546.9914703,
            "archived": false
        }
    ]
}
```

to work as follows:

1. as you said, reduced power of `mod` so it cannot move revisions into other `ID`s.  so a revision is only created upon modification, where it creates a full snapshot of the latest revision that got modified, and inserts it there with `.insert(0, ...)` so that the latest is 1st.
2. `nsapass timetravel ID TIME` will pick revision index `TIME` and make it current.  i.e. something like `tmp = db[ID].pop(TIME)` it, then `db[ID].insert(0, tmp)` it.
3. this revisioning will also store information about past deletions (aka archivings).  you can see that `ID` with `some other account` got deleted, hence having a revision with everything identical to previous one, except archived became `true`, and updated `timestamp`.
4. if `ID` `some other account` got deleted again, it will be fully wiped from database, including all of its past revisions.
5. here is one neat trick imo:  you see those spaces in `ID`?  e.g. `caveman protonmail`?  internally, `nsapass` to use `ID.split()`, then it will apply lazy matching for each word.  this way, we get to do things like `nsapass get c p` to mean identically to `nsapass get caveman protonmail`.  of course, `nsapass get c` to also work identically if no other `ID` started with `c`.

**Comment by int-e** [comment]
I think we should keep in mind that this is a (simple) password manager, not a revision control system. So IMO, the time travel feature is overkill. Keeping track of old ids also seems a bit too much to me, but I'm not feeling objective about that. On a positive note, maybe there's some value in having time stamps on entries--answeing the question of when they were added or last modified, which is the kind of information I might actually keep around in a note otherwise.

**Comment by al-caveman** [comment]
i agree.  complexity defeats the point since a passwords manager that is not readable in a few minutes (not auditable) is a huge problem by itself (i can't trust as such, and fortunately i don't have to).

but thing is, i already implemented it, and it seems pretty simple.  so here is the timetravel function:
```python
def timetravel(db, tags, tid):
    """travel into time tid"""
    printinfo('traveling through time...')
    tmp = db[tags].pop(tid)
    db[tags].insert(0, tmp)
    db[tmp[KEY_TAGS]] = db[tags]
    del db[tags]
```

and this is the part where it forgets past revisions:
```python
def forget(db, tags, retain):
    """forget the distant past of an entry from db"""
    db[tags] = [db[tags][0:retain]]
```

**Comment by int-e** [comment]
A minor observation: my fingers seem to keep typing `list` instead of `view`, how do you feel about renaming that command?

**Comment by al-caveman** [comment]
i agree.  `view` is historically so unnatural for *nix backgrounds.  better become back into `less` (i forgot why i made it `view`)

how about `ls`, and `rm` for `del`?
